connectors/memory/memory.go

```diff
@@ -245,9 +247,20 @@ func compareType(d1 dosa.FieldValue, d2 dosa.FieldValue) int8 {
 // Otherwise, search the partition for the exact same clustering keys. If there, fail
 // if not, then insert it at the right spot (sort.Search does most of the heavy lifting here)
 func (c *Connector) CreateIfNotExists(_ context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
-	return c.mergedInsert(ei.Def.Name, ei.Def.Key, values, func(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	err := c.mergedInsert(ei.Def.Name, ei.Def.Key, values, func(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) error {
 		return &dosa.ErrAlreadyExists{}
 	})
+	if err != nil {
```

Comment:
can this if statement be merged with line 252?
```
alreadyExistsFunc := func(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) error {
    return &dosa.ErrAlreadyExists{}
}
if err := c.mergedInsert(ei.Def.Name, ei.Def.Key, values, alreadyExistsFunc); err != nil {
  return err
}
```

you could probably even pull out the function definition (I'm not sure if the go-compiler is smart enough to realize it doesn't need to recreate that function everytime `CreateIfNotExists` is called)